fix: Allow i18n of form builder max/min character count. Earlier strings missed due to variable.

Anshumancanrock opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #22358 (GitHub issue number)
- Fixes CAL-6070(Linear issue number - should be visible at the bottom of the GitHub issue description)

**Key Changes:**
- Fixes missing German localization for form validation messages

## Visual Demo (For contributors especially)

#### Image Demo:

**Before (Issues):**

![68747470733a2f2f75706c6f6164732e6c696e6561722e6170702f65383662663935372d643832662d343635652d623230352d3133353535396634623632332f63626263323830322d633262662d346663362d393462362d3666383662656435366161372f626132666137](https://github.com/user-attachments/assets/b56aba0e-31f2-43fa-8825-c112bd24e07a)


- German users saw "Min. 20 characters required" instead of localized text

**After (Fixed):**

<img width="1918" height="937" alt="Screenshot 2025-07-11 233252" src="https://github.com/user-attachments/assets/39442dc6-0b75-4bb3-b49f-328c31090f1d" />


https://github.com/user-attachments/assets/3bec74ab-a382-4f15-9138-3cf08a0c1b85


- German users see: "Min. 20 Zeichen erforderlich"

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write **N/A** here and check the checkbox. **N/A** - These are internal URL construction fixes that don't affect public API.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

**Environment Setup:**
- No special environment variables required
- Standard Cal.com development setup

**Test Cases:**

 **German Translation Validation:**
   - Set browser language to German
   - Create/edit an event type with form fields requiring min/max characters
   - Verify validation messages appear in German: "Min. 20 Zeichen erforderlich"

**Expected Results:**
- German form validation messages display correctly

**Minimal Test Data:**

- German locale enabled in browser/system

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings
    
















<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Localized min/max character validation messages and wired schemas to use i18n with count placeholders, so German users now see correct error text.

- **Bug Fixes**
  - Replaced hardcoded length errors with translation keys and count interpolation; added locale strings.
  - Passed translateFn through booking responses and form-builder schemas to surface localized errors (CAL-6070).

<sup>Written for commit 89fdc33e186fa67b9689fa17d0e0d885507f54a7. Summary will update automatically on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

Anshumancanrock committed:
added german translation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=22367) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=22367) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=22367) it.</sub>

akashkinkarpandey commented:
@Anshumancanrock Can you please provide a loom video confirming the German translation ?

anikdhabal reviewed (dismissed):
Pls share the evidence of the fix

kart1ka committed:
Merge branch 'main' into german-lang-issue

kart1ka reviewed (changes requested):
does not seem to work

<img width="750" height="842" alt="Image" src="https://github.com/user-attachments/assets/b6f85bf9-faff-43ba-b5f4-6b73cc671e06" />

Anshumancanrock committed:
fixed count issue

Anshumancanrock committed:
fixed count issue

kart1ka commented:
@Anshumancanrock I don't see the video.

Anshumancanrock committed:
resolved type safety errors

kart1ka committed:
fix

kart1ka commented on packages/features/bookings/lib/getBookingResponsesSchema.ts:
```diff
@@ -317,7 +333,7 @@ function preprocess<T extends z.ZodType>({
 
         // Use fieldTypeConfig.propsType to validate for propsType=="text" or propsType=="select" as in those cases, the response would be a string.
         // If say we want to do special validation for 'address' that can be added to `fieldTypesSchemaMap`
-        if (["address", "text", "select", "number", "radio", "textarea"].includes(bookingField.type)) {
```
why was "textarea" removed?

Anshumancanrock replied:
You can add it sir ! It would still work without breaking anything. Please let me know if you want any changes.

kart1ka replied:
Pls make the changes if this was removed without a reason.

kart1ka committed:
Merge branch 'main' into german-lang-issue

kart1ka reviewed (commented):
Left a comment

Anshumancanrock committed:
Update getBookingResponsesSchema.ts

kart1ka reviewed (dismissed):
LGTM

Anshumancanrock committed:
Merge remote-tracking branch 'upstream/main' into german-lang-issue

Anshumancanrock committed:
fixes

Anshumancanrock committed:
removed 1 character case

Anshumancanrock commented:
@kart1ka Hey Sir ! Because of Rabbit AI suggestion i implemented "1 character" case without even thinking that the "1 character" case would never actually be displayed to users because of required field message.  So i reverted that change and now everything is working clean like before.

Here is the ss of translation after these changes:

<img width="1918" height="935" alt="Screenshot 2025-07-15 120921" src="https://github.com/user-attachments/assets/fdcc6985-c493-470b-91eb-83f2bf6e1ff3" />

Appreciate your time !

kart1ka reviewed (dismissed):
LGTM

Anshumancanrock commented:
@anikdhabal Please review sir and let me know if you want any changes. Appreciate your time !

kart1ka commented:
@Anshumancanrock It will be reviewed. It is going to take some time.

volnei committed:
Merge branch 'main' into german-lang-issue

kart1ka committed:
Merge branch 'main' into german-lang-issue

Anshumancanrock committed:
Merge branch 'calcom:main' into german-lang-issue

Anshumancanrock committed:
fixes

Anshumancanrock commented:
@keithwillcode  The "1 character" case will never show since required field validation runs first, but implemented the complete pattern to follow i18next best practices as per your suggestions.
 
 Here is the latest video after changes:


https://github.com/user-attachments/assets/30e2d172-1752-43e1-99f1-2a61d239385a

Please take a look and let me know if you’d like me to make any other changes.

Anshumancanrock commented:
@kart1ka @keithwillcode Pease review this pr sir . It's been open for a long time and I would appreciate your feedback when you have a moment.

anikdhabal committed:
Merge branch 'main' into german-lang-issue

anikdhabal committed:
Merge branch 'main' into german-lang-issue

kart1ka reviewed (changes requested):
left some comments

anikdhabal commented on apps/web/public/static/locales/de/common.json:
```diff
@@ -10,6 +10,10 @@
   "min_characters": "Min. Zeichen",
+  "min_characters_required_one": "Min. {{count}} Zeichen erforderlich",
+  "min_characters_required_other": "Min. {{count}} Zeichen erforderlich",
+  "max_characters_allowed_one": "Max. {{count}} Zeichen erlaubt",
+  "max_characters_allowed_other": "Max. {{count}} Zeichen erlaubt",
```
@Anshumancanrock if it's not use anywhere pls remove it

Anshumancanrock commented:
@kart1ka  The word Zeichen works for both ‘character’ and ‘characters’ in German. I am aware that min_characters_required_one is not used anywhere in the code, as it is overridden by the ‘This field is required’ message, but since this change was requested by @keithwillcode, I implemented it even though it is not used anywhere in the code.

Anshumancanrock committed:
fix

anikdhabal committed:
Merge branch 'main' into german-lang-issue

kart1ka reviewed (dismissed):
LGTM

kart1ka committed:
chore

kart1ka commented:
Build is successful now. Please review @volnei

volnei committed:
Merge branch 'main' into german-lang-issue

volnei committed:
Merge branch 'main' into german-lang-issue

volnei commented:
> @volnei could you please take a look when convenient? Thanks!

hey @Anshumancanrock, we don't need to introduce changes for specific languages, only english and let's lingo.dev to make appropriated translations for all supported languages.

Anshumancanrock committed:
fix

Anshumancanrock committed:
Merge branch 'main' into german-lang-issue

Anshumancanrock committed:
Merge branch 'main' into german-lang-issue

Anshumancanrock commented:
@kart1ka, please review this now.

dhairyashiil committed:
Merge branch 'main' into german-lang-issue

dhairyashiil reviewed (changes requested):
Everything looks good. Please resolve this last comment and let’s get this merged. I can see it’s been pending for a long time.
making it draft until then

Anshumancanrock committed:
type-check

dhairyashiil commented on packages/features/form-builder/schema.ts:
```diff
@@ -116,7 +116,7 @@ export const fieldTypesSchemaMap: Partial<
         isPartialSchema: boolean;
         ctx: z.RefinementCtx;
-        m: (key: string) => string;
+        m: (key: string, options?: any) => string;
```
this one?

Anshumancanrock committed:
type-fix

Anshumancanrock committed:
Merge branch 'main' into german-lang-issue

dhairyashiil reviewed (approved):
LGTM 👍🏼 Thank you for the contribution and for being so patient with all the back and forth communication, I know this can be tiring

dhairyashiil committed:
Merge branch 'main' into german-lang-issue

dhairyashiil committed:
Merge branch 'main' into german-lang-issue

anikdhabal committed:
Merge branch 'main' into german-lang-issue

pallava-joshi committed:
Merge branch 'main' into german-lang-issue

pallava-joshi committed:
Merge branch 'main' into german-lang-issue

emrysal reviewed (approved):
Looks good to me, but the title doesn't seem accurate, will update.

dhairyashiil committed:
Merge branch 'main' into german-lang-issue
